src/server/conn.rs

```diff
@@ -513,3 +553,165 @@ where
         }
     }
 }
+
+// TODO: different file? idk where to put this.
+mod rewindable {
+    use bytes::{Buf, BufMut, Bytes, IntoBuf};
+    use futures::{Async, Poll};
+    use std::io::{self, Read, Write};
+    use tokio_io::{AsyncRead, AsyncWrite};
+
+    #[derive(Debug)]
+    pub struct Rewindable<T>
+    where
+        T: AsyncRead + AsyncWrite,
+    {
+        pre: Option<Bytes>,
+        inner: T,
+    }
+
+    impl<T> Rewindable<T>
+    where
+        T: AsyncRead + AsyncWrite,
+    {
+        pub(super) fn new(tcp: T) -> Rewindable<T> {
+            Rewindable {
+                pre: None,
+                inner: tcp,
+            }
+        }
+        pub fn prepend(&mut self, bs: Bytes) {
+            self.pre.get_or_insert(bs);
+        }
+        pub fn into_inner(self) -> T {
+            self.inner
+        }
+    }
+
+    impl<T> Read for Rewindable<T>
+    where
+        T: AsyncRead + AsyncWrite,
+    {
+        #[inline]
+        fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+            if let Some(ref mut bs) = self.pre {
+                let pre = bs.clone().into_buf();
+                if pre.has_remaining() {
+                    let mut pre_reader = pre.reader();
+                    let pre_read_size = pre_reader.read(buf)?;
+                    bs.advance(pre_read_size);
+
+                    if pre_read_size < buf.len() {
+                        let (_, tail) = buf.split_at_mut(pre_read_size);
+                        let inner_read_size = self.inner.read(tail)?;
+
+                        return Ok(pre_read_size + inner_read_size);
+                    } else {
+                        return Ok(pre_read_size);
+                    }
+                }
+            }
+            return self.inner.read(buf);
+        }
+    }
+
+    impl<T> Write for Rewindable<T>
+    where
+        T: AsyncRead + AsyncWrite,
+    {
+        #[inline]
+        fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+            self.inner.write(buf)
+        }
+
+        #[inline]
+        fn flush(&mut self) -> io::Result<()> {
+            self.inner.flush()
+        }
+    }
+
+    impl<T> AsyncRead for Rewindable<T>
+    where
+        T: AsyncRead + AsyncWrite,
+    {
+        #[inline]
+        unsafe fn prepare_uninitialized_buffer(&self, buf: &mut [u8]) -> bool {
+            self.inner.prepare_uninitialized_buffer(buf)
+        }
+
+        #[inline]
+        fn read_buf<B: BufMut>(&mut self, buf: &mut B) -> Poll<usize, io::Error> {
+            if let Some(ref mut bs) = self.pre {
```

Comment:
Perhaps here and in `read`, the `pre` should be changed to `None` once all the bytes have been read. That will mean then skipping the rest of this `if` block on latter reads, and allow the `Bytes` to be freed since it's no longer needed.

Reply:
Absolutely, actually I'm not super confident on what is in here, but getting close to something that seems right.